samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemProducer.scala

```diff
@@ -161,72 +127,102 @@ class KafkaSystemProducer(systemName: String,
                                     envelope.getMessage.asInstanceOf[Array[Byte]])
 
     try {
-      sourceData.sendLock.synchronized {
-        val futureRef: Future[RecordMetadata] =
-          currentProducer.send(record, new Callback {
-            def onCompletion(metadata: RecordMetadata, exception: Exception): Unit = {
-              if (exception == null) {
-                //send was successful.
-                metrics.sendSuccess.inc
-              }
-              else {
-                error("Closing the producer because of an exception in callback: ", exception)
-                //If there is an exception in the callback, close producer.
-                closeAndNullifyCurrentProducer(currentProducer)
-
-                // we keep the exception and will throw the exception in the next producer.send()
-                // so the user can handle the exception and decide to fail or ignore
-                sourceData.exceptionInCallback.compareAndSet(
-                  null,
-                  new SamzaException("Unable to send message from %s to system %s." format(source, systemName),
-                    exception))
-
-                metrics.sendFailed.inc
-                error("Unable to send message on Topic:%s Partition:%s" format(topicName, partitionKey),
-                             exception)
-              }
-            }
-          })
-        sourceData.latestFuture = futureRef
-      }
+      currentProducer.send(record, new Callback {
+        def onCompletion(metadata: RecordMetadata, exception: Exception): Unit = {
+          if (exception == null) {
+            //send was successful.
+            metrics.sendSuccess.inc
+          }
+          else {
+            val producerException = new SystemProducerException(KafkaSystemProducer.FailedSendMsgFormat
+              .format(source, systemName, topicName, partitionKey), exception)
+
+            handleSendException(currentProducer, producerException, true)
+          }
+        }
+      })
       metrics.sends.inc
     } catch {
-      case e: Exception => {
-        error("Closing the producer because of an exception in send: ", e)
-
-        closeAndNullifyCurrentProducer(currentProducer)
+      case e : Exception =>
```

Comment:
Minor: s/e/originalException (clarifies that isFatalException is on originalException as opposed to producerException.)